### DISCUSSION TITLE
Ordered register fields

### DISCUSSION THREAD

**Discussion by LarsAsplund** [hdl-registers/hdl-registers#95] [Q&A] [answered]
Hi,

Let's say I have a register with a mix of field types and I have an 8-bit bit_vector. To support byte accesses and make register content more readable, I'd like to align that field to a byte boundary. Is there a way to do that? First I thought about adding "reserved" fields to create a gap but that only works well if the fields are ordered as they appear in the specification file. However, the different field types are grouped so that is not the case. Any ideas on how this can be solved? The only way I thought about is enforce a single field type, typically bit_vector, in registers where I have this requirement

**Comment by LarsAsplund** [comment]
The single field type workaround probably requires a recent Python version where dicts are ordered by default.

**Comment by LukasVik** [comment]
Yes, the fields are grouped with their type, unfortunate as that may be, due to the way we chose to organize the data file.

```
register.<type1>.<name1>.description = "..."
register.<type2>.<name2>.description = "..."
register.<type1>.<name3>.description = "..."
```

`name1` and `name3` will be placed first, and `name2` after them both. Within each type, however, fields should appear the way they appear in TOML.

This limitation is not present in the Python API, only in the data file parser. Updating the parser with a new format could be a solution, but that would be a very painful breaking change for a lot of people.

**Comment by LukasVik** [comment]
An alternative could be to pad each field type to a total width that is a multiple of eight. E.g. if you have two bits in your register, add six padding ones so that upcoming fields start at a multiple of eight. If you have a bit vector of width 12, add a padding one of width 4. And similar for enumeration and integer.

**Comment by LukasVik** [comment]
Just so that I understand the background,

1. What interface or programming language are you using where readability matters?
2. Does your project use an 8-bit register bus, or do you want to do sub-register accesses on a 32-bit bus?

**Reply by LarsAsplund** [reply]
Yes, in this project we can access individual bytes and half-words.

For readability, I was thinking about the hex dumps we sometimes need to decipher. Easier if a byte of information embedded in a wider word is byte aligned .

**Reply by LukasVik** [reply]
Ok I see. I agree that when looking manually at register values, it is very annoying that fields are not byte- or at least nibble-aligned. Looking forward I think this should be solved programmatically.

I plan to expand the generated Python code to give a nice way to read/write/print/values. For each register list it would generate a class that has named methods to read/write each register and each field, which would be useful for a Python-based system test environment. Would also have methods to print each register/field, or everything, in a nice formatted way showing native type values. It would take some "target accessor" object provided by the user as an argument. This object would be the proxy that accesses registers over SSH/telnet/JTAG or whatever the user is using, given a register address.

I think that would be cool.

**Comment by LukasVik** [comment]
hdl-registers is not written with sub-register accesses in mind. The register is seen as the smallest unit for access. The generated C++ class will do a 32-bit read-modify-write on "r_w" register fields, even if the field is byte aligned. Other generator projects place the "mode" setting on the field, hdl-registers places it on the register.

hdl-registers is also not written to support register widths other than 32. This is quite hard coded.

Both of these limitations are in place for simplicity and to keep complexity down. Most other generator projects suffer from scope creep, in my humble opinion, and seem to me quite bloated. I want to avoid that in this project, so hence I made some assumptions that come with limitations. I know that as a consequence, this project is not right for everyone.

With this said, I think support for non-32-bit register widths could be added quite cleanly and nicely. If that is the core problem you are facing. However, this would be a massive work at this moment. I can not do it unless some company chooses to sponsor this work.

**Reply by LarsAsplund** [reply]
Thanks for the clarification @LukasVik. Since I only use the parser and write my own generator I can probably work around the limitations with some extra naming conventions.

**Reply by LukasVik** [reply]
@LarsAsplund Have you started using the new generator API introduced in 5.0.0? https://hdl-registers.com/rst/extensions/extensions_custom_generator.html

It would be cool to hear from some third party who is actually using it. Any problems, ideas, things that are missing?

**Reply by LarsAsplund** [reply]
Yes, I have my own generator class derived from `RegisterCodeGenerator`. I don't have any issues with that class. I think the main thing right now is the balance between simplicity in the implementation and the level of control. In my case, there is already a documented register layout. While we intend to generate that document from the JSON/TOML specification in the end, it has already set an expectation on how things are ordered. We discussed the field ordering but now that I also looked closer at the register array, I see similar issues. For example, the document might say

reg x
reg y
n reserved words
a set of registers repeated 5 times
reg z

Ideally, I would like to put this in the same TOML/JSON spec but the spec is limited to one set of registers and one set of register arrays  and there is no way to interleave them, right? I guess I could create an extra level so that I get a spec of specs. A separate parser could deal with the top-level and then hdl-registers can manage the "sub specs". Any ideas on this?

**Reply by LukasVik** [reply]
Okay nice to hear that the new API is working well for you.

Yes I would agree that register array vs register ordering suffers from the same problem as we discussed for fields. Registers will be ordered in the way they appear in TOML, but always before all register arrays. Register arrays will be ordered in the way they appear in TOML, but always after all registers.

I know that shoehorning in hdl-registers usage in a project that used to have manual register handling is a pain. I've done once in a large project. But for me it was relatively simple, but still painful, because I could quickly do the corresponding SW changes also. So moving, renaming, reordering registers/arrays/fields was not an issue. However, I get that in a project where FPGA and SW are "far apart" it is impossible without an intermediary step where the hdl-registers layout is identical, or at least similar, to the previous manual layout.

Thinking about the area where we could theoretically allow manual handling:

1. Manual ordering of registers and register arrays
2. Manual ordering of fields
3. Manual index for register
4. Manual base index for field
5. Manual enum/integer field width
6. Manual enum encoding

I have said no to a previous request about manual register indexing, since I believe in principle if that is supported, there is no logical argument against supporting all the other manual handling as well. And I don't want that, since then hdl-registers would be just as bloated and complex as all the other tools. I think in this scenario it is my responsibility as maintainer to draw the line and remind what this project is about: _**generating**_ register code, not manually specifying it.

HOWEVER, with that said, I think 1 and 2 are different. They could come at no cost, and would not add complexity. And thinking about it more, I actually find them desirable. Associated fields should, for code readability, be placed next to each other, even if they're of different types.

So there is no ideological resistance from me, but there is a technical problem. We would need to change the data file format, and a lot of users use the old data file format. From:

```
register.<register_name>.<field_type>.<field_name>.description = "..."
register_array.<register_array_name>.register.<register_name>.description = "..."
```

to something like

```
<register_name>.register.<field_name>.<field_type>.description = "..."
<register_array_name>.register_array.<register_name>.register.description = "..."
```

But I think the parser could detect if the old or new format is used. For a while we could support both, with a deprecation warning about the old. I think it should be possible to write a script that converts automatically from the old format to the new. This would make user transition less painful.

**Reply by LarsAsplund** [reply, marked as answer]
Sounds good. For now, I proceeded with the discussed workarounds for field ordering and I extended our local JSON file to be a list of the standard specification objects at the top-level. First I read the JSON file with plain Python and then my code generator loops over these objects to generate the output. For each object I use your parser. This approach means that I can't build on the `RegisterCodeGenerator` class. Instead I work directly with the register list outputs from the parser. I'll be happy to test any updates in this area but for now I think we can close this discussion as "answered"
